### PR TITLE
UPBGE: Cull shadows when light frustum doesn't interstect with camera frustum

### PR DESCRIPTION
**Pull request by youle31**
Works for sun. Needs optimization

test file: http://pasteall.org/blend/index.php?id=44074
perf test file: http://pasteall.org/blend/index.php?id=44075

### CODE REVIEW

File: source/gameengine/Ketsji/KX_KetsjiEngine.cpp [hunk 1]
```diff
@@ -841,6 +841,17 @@ void KX_KetsjiEngine::RenderShadowBuffers(KX_Scene *scene)
...
 			raslight->HasShadowBuffer() && raslight->NeedShadowUpdate())
 		{
+			/* There was no culling for shadows other than if the object was outside light frustum.
+			 * We can't do a complete culling according to camera frustum because even if an object
+			 * is outside camera frustum, we should see his shadows. What we can do is check if light frustum
+			 * intersects with camera frustum. If this is not the case, we don't render shadows.
+			 */
+			MT_Vector3 lightfrustum[8];
+			raslight->GetShadowBox(lightfrustum);
+			int insideFrustum = KX_GetActiveScene()->GetActiveCamera()->BoxInsideFrustum(lightfrustum); // 0 = inside, 1 = intersect, 2 = outside
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_KetsjiEngine.cpp]: KX_GetActiveScene is not needed here, just replace by `scene`.

**Comment by youle31** [comment]
@lordloki : I can't see your comment anymore. I guess when we push --force the comments on old commits disappear because they are attached to the old commit. You can directly comment on pull request "conversation" or in "files changed" I think, like that the comments don't disappear.

About pyramid, I don't know how to write a function "PyramidInsideFrustum" but if you have an idea this would be better indeed.

**Comment by youle31** [comment]
Done for pyramid. I didn't test so I'll test. 

And panzergame told me something that I didn't understood:  "shadow are rendered in a limited area if we talk about depth, but the shadow is projected in a unlimited depth. So if the camera see an object aligned with the Z axis of the shadow box then the shadow worth update." so we need to discuss it.

**Comment by panzergame** [comment]
http://pasteall.org/pic/index.php?id=108185

The camera will see the shadow of the cube without see the shadow box.

**Comment by youle31** [comment]
The Shadow can be outside the light frustum? How can the Shadow be computed if it Is out light frustum?

**Comment by youle31** [comment]
Ok I understood. So do you think if we take a big value for clipend this would be ok?

**Comment by panzergame** [comment]
There's an issue with multiple cameras: currently the shadow are updated only one time at the begining of the render stage, but it will use only one camera for the culling and not all the cameras.

To solve this i will work on updating the shadow for each cameras but with a flag to avoid redundant update.

**Comment by youle31** [comment]
ah yes... right

**Comment by panzergame** [comment]
I tested the patch this morning and changed some things. I'm very surprised to not see any slowdown when the culling is not needed, when the camera perfectly saw the shadow.

About the pyramide shapes what about add an argument "count" to KX_Camera::BoxInsideFrustum to design the number of point and then in this function iterate of the same number of points.

**Comment by youle31** [comment]
I changed the pyramid and restored a box because even if it Is a bit less accurate we have only 8 points vs 8 plus 5 points if we have a pyramid plus an infinité box

File: source/gameengine/Rasterizer/RAS_OpenGLRasterizer/RAS_OpenGLLight.cpp [hunk 2]
```diff
@@ -200,6 +198,57 @@ MT_Matrix4x4 RAS_OpenGLLight::GetShadowMatrix()
 	return mat;
 }
 
+void RAS_OpenGLLight::GetShadowBox(MT_Vector3 *box)
+{
+	GPULamp *lamp;
+
+	if ((lamp = GetGPULamp())) {
```
Comment (panzergame) [inline source/gameengine/Rasterizer/RAS_OpenGLRasterizer/RAS_OpenGLLight.cpp]: i perfer avoid assignment in condition, instead do:
```
GPULamp *lamp = GetGPULamp();

if (lamp) {
```

File: source/gameengine/Rasterizer/RAS_OpenGLRasterizer/RAS_OpenGLLight.cpp [hunk 3]
```diff
@@ -200,6 +198,57 @@ MT_Matrix4x4 RAS_OpenGLLight::GetShadowMatrix()
...
+		if (light->GetLightData()->m_type == RAS_ILightObject::LIGHT_SUN) {
+
+			box[0][0] = box[1][0] = box[2][0] = box[3][0] = -GPU_lamp_frustum_size(lamp);
+			box[4][0] = box[5][0] = box[6][0] = box[7][0] = +GPU_lamp_frustum_size(lamp);
+			box[0][1] = box[1][1] = box[4][1] = box[5][1] = -GPU_lamp_frustum_size(lamp);
+			box[2][1] = box[3][1] = box[6][1] = box[7][1] = +GPU_lamp_frustum_size(lamp);
+			box[0][2] = box[3][2] = box[4][2] = box[7][2] = -FLT_MAX;
+			box[1][2] = box[2][2] = box[5][2] = box[6][2] = -GPU_lamp_clipstart(lamp);
+
+			MT_Matrix3x3 orientation = light->NodeGetWorldOrientation();
```
Comment (panzergame) [inline source/gameengine/Rasterizer/RAS_OpenGLRasterizer/RAS_OpenGLLight.cpp]: You could share the matrix/transform creation and apply to corner for sun light and spot light.

File: source/gameengine/Ketsji/KX_KetsjiEngine.cpp [hunk 4]
```diff
@@ -846,40 +862,56 @@ void KX_KetsjiEngine::RenderShadowBuffers(KX_Scene *scene)
 
 		raslight->Update();
 
-		if (light->GetVisible() && m_rasterizer->GetDrawingMode() == RAS_IRasterizer::RAS_TEXTURED &&
```
Comment (panzergame) [inline source/gameengine/Ketsji/KX_KetsjiEngine.cpp]: I will change this to use return statement instead of indentation monster.

Comment (youle31) [inline source/gameengine/Ketsji/KX_KetsjiEngine.cpp]: I didn't understood this. I let you fix if you don't mind. Thanks

File: source/gameengine/Ketsji/KX_KetsjiEngine.cpp [hunk 5]
```diff
@@ -898,10 +925,6 @@ void KX_KetsjiEngine::RenderFrame(KX_Scene *scene, KX_Camera *cam, unsigned shor
 
 	KX_SetActiveScene(scene);
 
-#ifdef WITH_PYTHON
```
Comment (youle31) [inline source/gameengine/Ketsji/KX_KetsjiEngine.cpp]: This has to be restored.... I don't know how... When I rebase this leads to weird things. This is why I prefer to make new branches hehe

**Comment by youle31** [comment]
Arf it doesn't work as expected. I'll try to draw frustums to see what is wrong... :/

**Pull request closed without merging**